File: python/tests/test_config.py
import os

import pytest
import yaml

from microservice_toolbox.config.loader import AppConfig, load_config

# ---- Deep Merge ----

def test_app_config_deep_merge():
    dst = {"a": 1, "b": {"c": 2}}
    src = {"b": {"d": 3}, "e": 4}

    AppConfig.deep_merge(dst, src)

    assert dst["a"] == 1
    assert dst["e"] == 4
    assert dst["b"]["c"] == 2
    assert dst["b"]["d"] == 3


# ---- Config Loading ----

def test_load_config_factory(tmp_path):
    """Verify the load_config() factory function works (Go LoadConfig parity)."""
    config_file = tmp_path / "factory.yaml"
    yaml.dump({"common": {"name": "factory-app"}}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("factory", input_args=[])
        assert isinstance(ac, AppConfig)
        assert ac.profile == "factory"
    finally:
        os.chdir(old_cwd)


def test_app_config_loading_and_addresses(tmp_path):
    """Verify loading, get_listen_addr, and get_grpc_listen_addr."""
    config_file = tmp_path / "test-profile.yaml"
    config_data = {
        "common": {"name": "test-app"},
        "capabilities": {
            "test-service": {
                "ip": "1.2.3.4", "port": "8080",
                "grpc_ip": "1.2.3.4", "grpc_port": "8081",
            }
        },
    }
    yaml.dump(config_data, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("test-profile", input_args=[])
        assert ac.profile == "test-profile"
        assert ac.get_listen_addr("test-service") == "1.2.3.4:8080"
        assert ac.get_grpc_listen_addr("test-service") == "1.2.3.4:8081"
    finally:
        os.chdir(old_cwd)


def test_app_config_missing_file():
    """Verify FileNotFoundError on missing profile (no bridge fallback)."""
    with pytest.raises(FileNotFoundError):
        load_config("non-existent-profile", input_args=[])


def test_grpc_missing_raises(tmp_path):
    """Verify get_grpc_listen_addr raises when grpc_ip/grpc_port are absent (Go parity)."""
    config_file = tmp_path / "no-grpc.yaml"
    yaml.dump({"capabilities": {"svc": {"ip": "1.2.3.4", "port": "8080"}}}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("no-grpc", input_args=[])
        with pytest.raises(ValueError):
            ac.get_grpc_listen_addr("svc")
    finally:
        os.chdir(old_cwd)


# ---- Decryption ----

def test_decrypt_secret_plaintext_passthrough(tmp_path):
    """Verify plaintext strings pass through unchanged."""
    config_file = tmp_path / "plain.yaml"
    yaml.dump({"common": {"name": "plain"}}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("plain", input_args=[])
        assert ac.decrypt_secret("normal_pass") == "normal_pass"
        assert ac.decrypt_secret("") == ""
        assert ac.decrypt_secret("ENC(no-closing-paren") == "ENC(no-closing-paren"
        assert ac.decrypt_secret("not-ENC(data)") == "not-ENC(data)"
    finally:
        os.chdir(old_cwd)


def test_decrypt_secret_enc_raises(tmp_path):
    """Verify ENC(...) block raises ValueError when decryption fails."""
    config_file = tmp_path / "secret.yaml"
    yaml.dump({"password": "ENC(dummy)"}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("secret", input_args=[])

        # Raw data is preserved
        assert ac.data["password"] == "ENC(dummy)"

        # Valid ENC() block but dummy content → must raise
        with pytest.raises(ValueError):
            ac.decrypt_secret("ENC(dummy)")
    finally:
        os.chdir(old_cwd)


# ---- Local Config ----

def test_get_local(tmp_path):
    """Verify get_local() returns values from the 'private' YAML section."""
    config_file = tmp_path / "local.yaml"
    yaml.dump({"private": {"local_setting": "value_xyz", "nested": {"a": 1}}}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("local", input_args=[])
        assert ac.get_local("local_setting") == "value_xyz"
        assert ac.get_local("nested") == {"a": 1}
        assert ac.get_local("missing") is None
    finally:
        os.chdir(old_cwd)


def test_get_local_empty(tmp_path):
    """Verify get_local() returns None when no private section exists."""
    config_file = tmp_path / "no-local.yaml"
    yaml.dump({"common": {"name": "test"}}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("no-local", input_args=[])
        assert ac.get_local("anything") is None
    finally:
        os.chdir(old_cwd)


def test_unmarshal_local(tmp_path):
    """Verify unmarshal_local() maps the private section to a class."""
    config_file = tmp_path / "unmarshal.yaml"
    yaml.dump({"private": {"local_setting": "value_xyz", "item_count": 5}}, open(config_file, "w"))

    class MyConfig:
        def __init__(self, local_setting=None, item_count=0):
            self.local_setting = local_setting
            self.item_count = item_count

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("unmarshal", input_args=[])
        
        # Test instantiating a type
        cfg = ac.unmarshal_local(MyConfig)
        assert cfg.local_setting == "value_xyz"
        assert cfg.item_count == 5

        # Test updating an existing instance
        cfg2 = MyConfig()
        ac.unmarshal_local(cfg2)
        assert cfg2.local_setting == "value_xyz"
        assert cfg2.item_count == 5
    finally:
        os.chdir(old_cwd)


# ---- Set Logger ----

def test_set_logger(tmp_path):
    """Verify set_logger replaces the logger safely."""
    config_file = tmp_path / "logger.yaml"
    yaml.dump({"common": {"name": "logger-test"}}, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("logger", input_args=[])
        old_logger = ac.logger

        # Setting None should not crash (ensure_safe_logger wraps it)
        ac.set_logger(None)
        assert ac.logger is not None
        assert ac.logger is not old_logger
    finally:
        os.chdir(old_cwd)


# ---- CLI Override Scope ----

def test_cli_override_targets_single_capability(tmp_path):
    """Verify CLI --host/--port overrides only the target capability (Go parity)."""
    config_file = tmp_path / "cli.yaml"
    yaml.dump({
        "common": {"name": "my-svc"},
        "capabilities": {
            "my-svc": {"ip": "0.0.0.0", "port": "9000"},
            "other-svc": {"ip": "0.0.0.0", "port": "9001"},
        },
    }, open(config_file, "w"))

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("cli", input_args=["--host", "10.0.0.1", "--port", "5555"])

        # Target capability (my-svc from common.name) should be overridden
        assert ac.get_listen_addr("my-svc") == "10.0.0.1:5555"

        # Other capability should NOT be affected
        assert ac.get_listen_addr("other-svc") == "0.0.0.0:9001"
    finally:
        os.chdir(old_cwd)


# ---- Env Expansion ----

def test_env_expansion(tmp_path):
    """Verify ${VAR:default} expansion in YAML loading."""
    config_file = tmp_path / "env.yaml"
    yaml.dump({
        "private": {
            "host": "${TEST_HOST:localhost}",
            "port": "${TEST_PORT:8080}",
        }
    }, open(config_file, "w"))

    os.environ["TEST_HOST"] = "127.0.0.5"
    if "TEST_PORT" in os.environ:
        del os.environ["TEST_PORT"]

    old_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        ac = load_config("env", input_args=[])
        assert ac.get_local("host") == "127.0.0.5"
        # YAML parses 8080 as int unless quoted
        assert int(ac.get_local("port")) == 8080
    finally:
        os.chdir(old_cwd)

File: cpp/tests/test_config.cpp
#include "../include/microservice_toolbox/config/AppConfig.hpp"
#include <cassert>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <unistd.h>

using namespace microservice_toolbox::config;

// Helper to ensure config directory exists
void ensure_config_dir() {
#ifdef _WIN32
  system("mkdir config 2>nul");
#else
  system("mkdir -p config");
#endif
}

void test_load_config_factory() {
  std::cout << "Testing LoadConfig Factory..." << std::endl;
  auto ac =
      LoadConfigWithLogger("standalone", std::make_shared<StdOutLogger>());
  assert(ac != nullptr);
  assert(ac->GetProfile() == "standalone");
  std::cout << "  Passed." << std::endl;
}

void test_address_resolution() {
  std::cout << "Testing Address Resolution..." << std::endl;
  ensure_config_dir();

  std::ofstream ofs("config/test.yaml");
  ofs << "common:\n  name: test-app\n"
      << "capabilities:\n"
      << "  log_server: {ip: '127.0.0.2', port: '9999'}\n"
      << "  config_server: {ip: '127.0.0.2', port: '9999'}\n"
      << "  svc:\n    ip: 127.0.0.2\n    port: '8080'\n    grpc_ip: "
         "127.0.0.2\n    grpc_port: '8081'";
  ofs.close();

  auto ac = LoadConfig("test");

  assert(ac->GetListenAddr("svc") == "127.0.0.2:8080");
  assert(ac->GetGRPCListenAddr("svc") == "127.0.0.2:8081");

  std::remove("config/test.yaml");
  std::cout << "  Passed." << std::endl;
}

void test_decrypt_secret_logic() {
  std::cout << "Testing DecryptSecret Hardened Logic..." << std::endl;
  auto ac = LoadConfig("standalone");

  assert(ac->DecryptSecret("normal_pass") == "normal_pass");
  assert(ac->DecryptSecret("") == "");
  assert(ac->DecryptSecret("ENC(no-close") == "ENC(no-close");
  assert(ac->DecryptSecret("not-ENC(data)") == "not-ENC(data)");

  try {
    std::string decrypted = ac->DecryptSecret("ENC(dummy)");
    // If we reach here, bridge somehow succeeded (unlikely in test)
    assert(!decrypted.empty());
  } catch (const std::exception& e) {
    std::string err = e.what();
    assert(!err.empty());
    std::cout << "  Passed (Caught Transparent Error: " << err << ")" << std::endl;
  }
}

void test_get_local() {
  std::cout << "Testing GetLocal..." << std::endl;
  ensure_config_dir();

  std::ofstream ofs("config/standalone.yaml");
  ofs << "common: {name: test}\n"
      << "local:\n  local_setting: value_xyz\n  nested:\n    val: 123\n    "
         "key: nested_value";
  ofs.close();

  auto ac = LoadConfig("standalone");
  if (ac->GetLocal("local_setting") != "value_xyz") {
    std::cerr << "  FAILED: Expected 'value_xyz', got '"
              << ac->GetLocal("local_setting") << "'" << std::endl;
    assert(false);
  }
  if (ac->GetLocal("nested.val") != "123") {
    std::cerr << "  FAILED: Expected '123', got '" << ac->GetLocal("nested.val")
              << "'" << std::endl;
    assert(false);
  }
  if (ac->GetLocal("nested.key") != "nested_value") {
    std::cerr << "  FAILED: Expected 'nested_value', got '"
              << ac->GetLocal("nested.key") << "'" << std::endl;
    assert(false);
  }
  assert(ac->GetLocal("missing") == "");

  std::remove("config/standalone.yaml");
  std::cout << "  Passed." << std::endl;
}

void test_grpc_missing_error() {
  std::cout << "Testing gRPC Missing Error (Hardened)..." << std::endl;
  ensure_config_dir();

  std::ofstream ofs("config/test.yaml");
  ofs << "common: {name: test}\n"
      << "capabilities:\n"
      << "  log_server: {ip: '127.0.0.2', port: '9999'}\n"
      << "  config_server: {ip: '127.0.0.2', port: '9999'}\n"
      << "  svc:\n    ip: 127.0.0.2\n    port: '8080'";
  ofs.close();

  auto ac = LoadConfig("test");

  bool threw = false;
  try {
    ac->GetGRPCListenAddr("svc");
  } catch (const std::runtime_error &) {
    threw = true;
  }
  assert(threw == true);

  std::remove("config/test.yaml");
  std::cout << "  Passed." << std::endl;
}

void test_env_expansion() {
  std::cout << "Testing Environment Variable Expansion..." << std::endl;

  // Using 'standalone' profile because it's known by the engine,
  // but we use a local override to test expansion.
  std::ofstream ofs("standalone.yaml");
  ofs << "common: {name: expansion-test}\n"
      << "local:\n"
      << "  host: ${TEST_HOST:localhost}\n"
      << "  port: ${TEST_PORT:8080}\n";
  ofs.close();

  setenv("TEST_HOST", "127.0.0.5", 1);
  unsetenv("TEST_PORT");

  auto ac = LoadConfig("standalone");

  if (ac->GetLocal("host") != "127.0.0.5") {
    throw std::runtime_error("Env expansion failed for TEST_HOST. Got: " +
                             ac->GetLocal("host"));
  }
  if (ac->GetLocal("port") != "8080") {
    throw std::runtime_error(
        "Env expansion default failed for TEST_PORT. Got: " +
        ac->GetLocal("port"));
  }

  std::cout << "  Passed." << std::endl;
  std::remove("standalone.yaml");
}

void test_mirror_integrity() {
  std::cout << "Testing Mirror Integrity (Full Sync)..." << std::endl;
  ensure_config_dir();

  std::ofstream ofs("config/test.yaml");
  ofs << "common: {name: mirror-test}\n"
      << "capabilities:\n"
      << "  sync_svc: {ip: '10.0.0.1', port: '1234'}";
  ofs.close();

  auto ac = LoadConfig("test");

  // We can't access data_ directly (it's local), but we can verify the sync log
  // and check if lookups work.
  assert(ac->GetListenAddr("sync_svc") == "10.0.0.1:1234");

  std::cout << "  Passed (In-memory mirror verified)." << std::endl;
  std::remove("config/test.yaml");
}

int main() {
  // Cleanup any interference
  std::remove("test_config.yaml");

  try {
    test_load_config_factory();
    test_address_resolution();
    test_decrypt_secret_logic();
    test_get_local();
    test_grpc_missing_error();
    test_env_expansion();
    test_mirror_integrity();

    std::cout << "\n=======================================" << std::endl;
    std::cout << "  All C++ Toolbox Parity Tests Passed!" << std::endl;
    std::cout << "=======================================" << std::endl;
  } catch (const std::exception &e) {
    std::cerr << "\n!!! TEST FAILED: " << e.what() << std::endl;
    return 1;
  }
  return 0;
}
